java-lib/src/main/java/com/yammer/metrics/core/WavefrontHistogram.java

```diff
@@ -91,18 +96,30 @@ public void update(int value) {
   }
 
   /**
-   * Helper to retrieve the current bin. Assumes proper synchronization.
+   * Helper to retrieve the current bin. Will be invoked per thread.
    */
   private MinuteBin getCurrent() {
-    long minMillis = minMillis();
+    long currMinMillis = minMillis();
+
+    long key = Thread.currentThread().getId();
+    LinkedList<MinuteBin> bins;
+    if (map.containsKey(key)) {
+      bins = map.get(key);
+    } else {
+      bins = new LinkedList<>();
```

Comment:
this code is problematic, given that this can be called concurrently, you need to use putIfAbsent() and check the result to make sure another thread isn't also doing the same thing